File: joined.py
from collections import deque
from math import ceil
import sys


class Node:
    def __init__(self, i, w):
        self.index = i
        self.word = w
        self.visited = False
        self.distance = 100000
        self.next_node = None


def main():
    #get words to join
    try:
        str_start = sys.argv[1]
        str_end = sys.argv[2]
    except:
        print("please give two words to join in args")
        exit(1)

    dictionary = []
    nodes = []

    #put search words at the start to make the search slightly faster
    dictionary.append(str_start)
    dictionary.append(str_end)

    try:
        #read in dictionary
        for line in sys.stdin :
            dictionary.append(line.rstrip())
    except:
        print("please pass a dictionary to stdin")
        exit(1)
    
    #remove duplicates
    dictionary = list(dict.fromkeys(dictionary))
    for i, word in enumerate(dictionary) :
        nodes.append(Node(i, word))

    #finding where the start and end words are in the nodes list
    start = nodes[dictionary.index(str_start)]
    end = nodes[dictionary.index(str_end)]

    #search for links in singled and double joined mode
    single_len = bfs(nodes, dictionary, True)
    double_len = bfs(nodes, dictionary, False)

    #print results
    print(start.word, end.word)
    print(*single_len)
    print(*double_len)

    exit(0)


def bfs(nodes, dictionary, single):
    start = nodes[0]
    end = nodes[1]

    #visit the start node
    start.distance = 0
    start.visited = True

    #queue the starting node and start the search
    queue = deque([start])
    #while stack still has items
    while(queue): 
        node = queue.popleft()

        neighbours = find_neighbours(nodes, node, single)
        print(neighbours)

        for node in neighbours:
            #check if we reached the end word
            if(node == end):
                #create a list of words joining start and end to return
                out = [node.distance, node.word]
                while not node.next_node == None:
                    node = node.next_node
                    out.insert(1, node.word)
                return out
            else:
                queue.append(node)

    #didn't find a link
    return[0]


def find_neighbours(nodes, left, single):
    #minimum length of the matching part of the left word
    suffix_len = int(ceil(len(left.word))/2)

    neighbours = []

    for right in nodes :
        if(right.visited):
            continue

        #check how big the matching string must be
        preffix_len = int(ceil(len(right.word))/2)

        match_len = min(suffix_len, preffix_len) if single else max(suffix_len, preffix_len) 

        #check if suffix matches prefix of next word
        if(left.word[-match_len:] == right.word[0:match_len]):
            #visit node
            right.next_node = left
            right.distance = left.distance + 1
            right.visited = True
            #save the neighbour to the return list
            neighbours.append(right)
        
        return neighbours


if __name__ == "__main__":
    main()

File: Joined.java
import java.util.Collections;
import java.util.LinkedHashSet;
import java.util.LinkedList;
import java.util.List;
import java.util.Scanner;
import java.util.Set;

public class Joined {  
    public static String[] dictionary;
    public static Node[] graph;
    public static String start;
    public static String end;

    public static void main(String[] args) {
        try {
            start = args[0];
            end = args[1];
        } catch(IndexOutOfBoundsException e) {
            System.out.println("please give two words to join in args");
            System.exit(1);
        }

        dictionary = loadDict();
        graph = graphFromDict();
        

        System.out.println(start + " " + end);

        if (start == end) {
            System.out.println("1 " + start + " " + end);
            System.out.println("1 " + start + " " + end);
        }
    }

    private static Node[] graphFromDict() {
        graph = new Node[dictionary.length];
        
        for(int i = 0; i < graph.length; i++) {
            graph[i] = new Node(i, dictionary[i]);
        }
        return graph;
    }

    /**
     * Reads in the dictionary words from standard input
     * @return sorted dictionary as an array
     */
    public static String[] loadDict() {
        // set to avoid duplicates
        Set<String> temp = new LinkedHashSet<String>();
        Scanner sc = new Scanner(System.in);
        if(!sc.hasNext()) {
            System.out.println("please pass a dictionary to standard input");
            System.exit(1);
        }

        do {
            temp.add(sc.nextLine().strip().toLowerCase());
        } while (sc.hasNextLine());
        sc.close();
    
        // make sure target word is in the dictionary by adding it
        temp.add(end);

        // convert dictionary to list so it can be sorted alphabetically
        List<String> temp2 = new LinkedList<String>(temp);
        Collections.sort(temp2);
        return (String[]) temp2.toArray();
    }

    public static class Node{
        int index;
        String word;
        boolean visited = false;
        int distance;
        Node root;
        int length;
        int half;

        Node(int i, String w) {
            index = i;
            word = w;
            length = word.length();
            half = (w.length() + 1) / 2;
            distance = 2147483647; // signed integer limit
        }
    }
}
